### ISSUE TITLE
Any chance we could get Claude.ai too?

### ISSUE THREAD

**Issue by gobijan** [yaroslavyaroslav/OpenAI-sublime-text#66]
I slowly made the switch from OpenAI to [Claude.ai](https://claude.ai/new) as the Sonnet Model is super fast and top scorer in benchmarks. Are you planning on extending the providers to work with other non OpenAIish APIs?

**Comment by yaroslavyaroslav** [comment]
Yeah, there's a problem with that. 

Both claude and gemeni API require strict order of messages to be sent, like `[user, assistant, user, assistant]`, which is not the case for openai API thus for this plugin. This is the main blocker to add support for them here, bc this plugin treats each text chunk as a separate message, so it's like `[user, user, user, assistant]`.

In addition to that I'm pretty sure they have made functions support (coming pretty soon) they way as well. 

So in overall it's kinda pretty complex adapter required to communicate with those, and I have no plans to implement it by my own for the time being.

**Comment by gobijan** [comment]
Yeah totally get your point. It would be a lot of work. 

Do you think that something like https://github.com/SilasMarvin/lsp-ai/ could be of any help?

**Comment by jalsco** [comment]
nice discussion - I also like the idea of Claude engine over the chatGPT one.
But understand that you have other priorities 😥

**Comment by yaroslavyaroslav** [comment]
Actually things got changed recently, especially after their introduction of [MCP](https://www.anthropic.com/news/model-context-protocol), so i'm about to add support of their API to this plugin. Not sure how soon though.

**Comment by jalsco** [comment]
That is awesome ! Thank you 👍

I live in Philippines for most of each year, and from this location I have had problems add extra funds into Anthropic account, so no urgency from my side. In the meantime (today) I will be trying to use your plugin with Grok 2. I am expecting that to be quite straightforward as Grok's documentation says it can just use the OpenAI interface unchanged (just pointing it to Grok and supply the Grok API KEY ) .   I hope that works and thank you so much for developing this 🙏🏻🙏🏻🙏🏻

About Grok - it is much cheaper - and is fast catching up to the others. It is still a long way behind ChatGPT and Claude for coding support BUT there is a massive upgrade to Grok3 scheduled for "December 2024" so it is possible it will catch up to the others within a few weeks. Maybe.... 🤷🏻‍♂️

**Comment by jalsco** [comment]
Well maybe not so easy to just run Grok via a plugin designed for OpenAI, not matter what X/Ai say ...

I got the plugin setup - and customized - clean as a whistle, no errors or messages (though my machine is on py 3.13 ).  One issue if the super-k keystroke..... in my Windows10  that is taken by some system level "project to wireless display" service.  So i changed it to super-o (mostly because o is near k ). 

Anyway - I ctrl-shift-p and enter: OpenAI: Chat Model Select. 
![image](https://github.com/user-attachments/assets/57707ab4-7b19-4d8a-89d2-dd22521ff374)

And I just to run a a prompt thru to Grok and get a reply.... 

So I select one of the "panel" ones. 

It does drop me down to the areas to type in a question/prompt. Which I do ...   and then it, sadly, it bombs out with some py errors.  

```
ERROR:root:OpenAI error: The requested resource was not found. Please check the URL and try again. Documentation is available at https://docs.x.ai/
Traceback (most recent call last):
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 301, in handle_response
    self.handle_chat_response()
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 281, in handle_chat_response
    response: HTTPResponse | None = self.provider.execute_response()
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_network_client.py", line 114, in execute_response
    return self._execute_network_request()
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_network_client.py", line 134, in _execute_network_request
    raise UnknownException(error_data.get('error').get('message'))
OpenAI completion.plugins.errors.OpenAIException.UnknownException: The requested resource was not found. Please check the URL and try again. Documentation is available at https://docs.x.ai/
error: OpenAI error
The requested resource was not found. Please check the URL and try again. Documentation is available at https://docs.x.ai/
ERROR:root:OpenAI error: The requested resource was not found. Please check the URL and try again. Documentation is available at https://docs.x.ai/
Traceback (most recent call last):
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 301, in handle_response
    self.handle_chat_response()
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 281, in handle_chat_response
    response: HTTPResponse | None = self.provider.execute_response()
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_network_client.py", line 114, in execute_response
    return self._execute_network_request()
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_network_client.py", line 134, in _execute_network_request
    raise UnknownException(error_data.get('error').get('message'))
OpenAI completion.plugins.errors.OpenAIException.UnknownException: The requested resource was not found. Please check the URL and try again. Documentation is available at https://docs.x.ai/
error: OpenAI error
The requested resource was not found. Please check the URL and try again. Documentation is available at https://docs.x.ai/
```

**Comment by jalsco** [comment]
I tried a few things for those errors above. Restarted Sublime, and tried ONLY `Example | panel | grok-beta` , and gave it one easy prompt... something like "Please tell me about how Grok compares to the other more well-known LLMs ?" . These are the errors: 

```
Exception in thread Thread-41:
Traceback (most recent call last):
  File "./python3.8/threading.py", line 932, in _bootstrap_inner
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 490, in run
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 403, in manage_chat_completion
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 301, in handle_response
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_worker.py", line 281, in handle_chat_response
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_network_client.py", line 114, in execute_response
  File "C:\Users\Sean\AppData\Roaming\Sublime Text 3\Installed Packages\OpenAI completion.sublime-package\plugins/openai_network_client.py", line 129, in _execute_network_request
AttributeError: 'str' object has no attribute 'get'
```

**Comment by yaroslavyaroslav** [comment]
@jalsco Text in first message seems to say that there's something wrong with the url you're trying to request.  I suggest you to create a separate issue for that and profile the full plugin settings there.

**Comment by jalsco** [comment]
Am I supposed to `pip intall openai` as a prerequisite ? 

I have not done that , as nowhere did it say i need to...

**Comment by yaroslavyaroslav** [comment]
Thanks for pinging me. Here's a little update about the task ongoing progress.

tldr; 
It's none 😀

Ok, if to be more specific here's the thing:

1. Task is now to be implemented in that repo https://github.com/yaroslavyaroslav/llm_runner/issues/1 since Python code has nothing to do with any backend now.
2. On Rust's side, things got a lot better when it comes to different APIs support:
    3. Each network type is isolated from one another, so you don't have to untangle any logic of how things are to make the type.
    4. Strict messages ordering has been implemented recently, so this can be considered as a foundation to compose all the consequent messages of a single role effortlessly — the main struggle to bring DeepSeek support as I imagine it. Same thing it can be made in isolated way to not mess with the rest of the API's or lib parts.

Now the bad things comes:

1. I run completely out of time for the next few months for sure, and when I will have time, I'll spend it on migrating OpenAI's `/completions` gateway to their new `request` or whatever, which brings internet search support on the table and then to tune up a model picker workflow a little bit, since I think I found a little bit better flow of models and modes switching. So it won't be Anthropic what I would invest my time in; for DeepSeek, there's a little more chance a task to be solved though.
2. Package Control is quite a down time for a month for the time being, so no updates are received to ST users since then, and it's uncertain when it'd be fixed. So there's no rush for me to rush this release as well, since adoption would be minimal.
3. All of the features made so far within this project are a subject of my personal need. So to be honest, I feel none for Anthropic models. I mean, I heard that they're better, but the one that I've played with in zed isn't impressed me that much that it convinces me to switch to them. OpenAI brand debacle bothers me for sure, but not that much to spend like a few days of my work on it.

So no conclusion out of this, just sharing some thoughts. 

However, if some of you folks are keen on Anthropic or DeepSeek and want to contribute here, I'd really do my best to onboard them into llm_runner code, so feel free to volunteer in this issue: https://github.com/yaroslavyaroslav/llm_runner/issues/1. As I said, just yet the code is quite isolated and clean in this part.

**Comment by yaroslavyaroslav** [comment]
So after two years of waiting here you are guys. 6.0.0 should just work.

**Comment by gobijan** [comment]
Nice! Time for a rename? :D

**Comment by yaroslavyaroslav** [comment]
If there is a chance that this repo repeats clawdbot->openclawd journey then I'm down for it 😀.
